ports/esp32/machine_pwm.c

```diff
@@ -113,320 +79,470 @@ static bool pwm_inited = false;
-    }
-    int max_duty = (1 << timer.duty_resolution) - 1;
-    if (channel_duty < 0) {
-        channel_duty = 0;
-    } else if (channel_duty > max_duty) {
-        channel_duty = max_duty;
-    }
-    check_esp_err(ledc_set_duty(self->mode, self->channel, channel_duty));
-    check_esp_err(ledc_update_duty(self->mode, self->channel));
-
-    /*
-    // Bug: Sometimes duty is not set right now.
-    // Not a bug. It's a feature. The duty is applied at the beginning of the next signal period.
-    // Bug: It has been experimentally established that the duty is set during 2 signal periods, but 1 period is expected.
-    // See https://github.com/espressif/esp-idf/issues/7288
-    if (duty != get_duty_u16(self)) {
-        PWM_DBG("set_duty_u16(%u), get_duty_u16():%u, channel_duty:%d, duty_resolution:%d, freq_hz:%d", duty, get_duty_u16(self), channel_duty, timer.duty_resolution, timer.freq_hz);
-        esp_rom_delay_us(2 * 1000000 / timer.freq_hz);
-        if (duty != get_duty_u16(self)) {
-            PWM_DBG("set_duty_u16(%u), get_duty_u16():%u, channel_duty:%d, duty_resolution:%d, freq_hz:%d", duty, get_duty_u16(self), channel_duty, timer.duty_resolution, timer.freq_hz);
-        }
+    if ((duty < 0) || (duty > UI16_DUTY)) {
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("duty_u16 must be from 0 to %d"), UI16_DUTY);
     }
-    */
 
-    self->duty_x = HIGHEST_PWM_RES;
-    self->duty_u16 = duty;
+    self->duty_x = UI_RES_16_BIT;
+    self->duty_ui = duty;
+    apply_duty(self);
 }
 
 static void set_duty_u10(machine_pwm_obj_t *self, int duty) {
-    pwm_is_active(self);
-    if ((duty < 0) || (duty > MAX_DUTY_U10)) {
-        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("duty must be from 0 to %u"), MAX_DUTY_U10);
+    if ((duty < 0) || (duty > UI10_DUTY)) {
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("duty must be from 0 to %u"), UI10_DUTY);
     }
-    set_duty_u16(self, duty << (UI_RES_16_BIT - PWM_RES_10_BIT));
-    self->duty_x = PWM_RES_10_BIT;
-    self->duty_u10 = duty;
+    self->duty_x = UI_RES_10_BIT;
+    self->duty_ui = duty;
+    apply_duty(self);
 }
 
 static void set_duty_ns(machine_pwm_obj_t *self, int ns) {
-    pwm_is_active(self);
-    if ((ns < 0) || (ns > duty_to_ns(self, UI_MAX_DUTY))) {
-        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("duty_ns must be from 0 to %d ns"), duty_to_ns(self, UI_MAX_DUTY));
+    if ((ns < 0) || (ns > duty_to_ns(self, UI16_DUTY))) {
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("duty_ns must be from 0 to %d ns"), duty_to_ns(self, UI16_DUTY));
     }
-    set_duty_u16(self, ns_to_duty(self, ns));
-    self->duty_x = -HIGHEST_PWM_RES;
-    self->duty_ns = ns;
+    self->duty_x = -UI_RES_16_BIT;
+    self->duty_ui = ns;
+    apply_duty(self);
 }
 
-/******************************************************************************/
+static void set_duty(machine_pwm_obj_t *self) {
```

Comment:
The control flow `set_duty(self) -> set_duty_xyz(self, arg) -> apply_duty(self)` seems a bit confusing. Do you think there is another way this function could be structured, perhaps:

```c
verify_duty(self); // Checks self->duty_x and self->duty_ui are in range, raises if out of range

// do other PWM setup

apply_duty(self)
```

It still seems sub-optimal that `self` is updated even if the arguments are invalid, but I can't see an easy way around that.